src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java

```diff
@@ -285,8 +290,85 @@ protected void retrieve(@CheckForNull SCMSourceCriteria criteria, SCMHeadObserve
                                " Check the configuration before running this job again.");
                 return;
             }
-            getFullyInitializedGitSCMSource().accessibleRetrieve(criteria, observer, event, listener);
+
+            if (hasLegacyTraits()) {
```

Comment:
Even there are no "legacy" discover traits enabled, the `GitSCMSource#retrieve` method will still clone the repo but will just not attempt to go through branches. So to help with performance, I've added this check to not bother running the legacy retrieve if there are not applicable traits anyway.